### PR TITLE
 Testing: Upgrade React and Enzyme to the latest version

### PR DESCRIPTION
**Pull request by gziolo**
## Description
Another attempt to upgrade React and Enzyme to the latest version. We started exploration in #2813 and this PR should finally make it happen. All tests work as expected.

As part of this diff I also improved the behavior of throwing an error when test executes `console.error`. I provided new logic which uses spies and `expect` call to validate if test should fail. New output looks as follow:

>  ● InserterMenu › should disable already used blocks with `usedOnce`
> 
>     expect(jest.fn()).not.toHaveBeenCalled()
>     
>     Expected mock function not to be called but it was called with:
>       ["Warning: Expected `onMouseLeave` listener to be a function, instead got a value of `boolean` type.
>         in button (created by InserterMenu)
>         in div (created by InserterMenu)
>         in div (created by InserterMenu)
>         in div (created by InserterMenu)
>         in div (created by InserterMenu)
>         in InserterMenu (created by WrapperComponent)
>         in WrapperComponent"], ["Warning: Expected `onMouseEnter` listener to be a function, instead got a value of `boolean` type.
>         in button (created by InserterMenu)
>         in div (created by InserterMenu)
>         in div (created by InserterMenu)
>         in div (created by InserterMenu)
>         in div (created by InserterMenu)
>         in InserterMenu (created by WrapperComponent)
>         in WrapperComponent"]
>       
>       at Object.<anonymous> (test/setup-test-framework.js:23:23)
>       at process._tickCallback (internal/process/next_tick.js:109:7)

It's much easier to read and it properly outputs all warnings and errors.

There are still 2 skipped tests. I will take a closer look at them later and decide if we can fix them or they should be removed.

## How Has This Been Tested?
Travis should be happy.

Run `npm test` locally.

To test new behavior around `console.error` and `console.warn` simply put such call in one of the tests.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

### CODE REVIEW

File: components/navigable-menu/test/index.js [hunk 1]
```diff
@@ -18,6 +18,7 @@ const { UP, DOWN } = keycodes;
 describe( 'NavigableMenu', () => {
 	// Skipping this this because the `isVisible` check in utils/focus/tabbable.js always returns false in tests
 	// Probbably a jsdom issue
+	// eslint-disable-next-line jest/no-disabled-tests
 	it.skip( 'should navigate by keypresses', () => {
```
Comment (gziolo) [inline components/navigable-menu/test/index.js]: I will look into it later. I added this eslint rule to hide warning.

File: editor/inserter/menu.js [hunk 2]
```diff
@@ -290,8 +290,8 @@ export class InserterMenu extends Component {
 				onClick={ this.selectBlock( block.name ) }
 				ref={ this.bindReferenceNode( block.name ) }
 				tabIndex="-1"
-				onMouseEnter={ ! disabled && this.props.showInsertionPoint }
-				onMouseLeave={ ! disabled && this.props.hideInsertionPoint }
+				onMouseEnter={ ! disabled ? this.props.showInsertionPoint : null }
```
Comment (gziolo) [inline editor/inserter/menu.js]: This change fixes warnings in the InserterMenu component:
> Warning: Expected `onMouseLeave` listener to be a function, instead got a value of `boolean` type.
> Warning: Expected `onMouseEnter` listener to be a function, instead got a value of `boolean` type.

File: test/setup-test-framework.js [hunk 3]
```diff
@@ -1,34 +1,25 @@
...
+	spyWarn.mockReset();
+} );
 
-		return renderToStaticMarkup( element );
-	},
-} ) );
+afterEach( () => {
+	expect( spyError ).not.toHaveBeenCalled();
+	expect( spyWarn ).not.toHaveBeenCalled();
+} );
```
Comment (youknowriad) [inline test/setup-test-framework.js]: ❤️

File: package.json [hunk 4]
```diff
@@ -28,12 +28,12 @@
     "moment-timezone": "0.5.13",
     "mousetrap": "1.6.1",
     "prop-types": "15.5.10",
-    "react": "15.6.1",
+    "react": "16.0.0",
```
Comment (youknowriad) [inline package.json]: 🎉

**Review by youknowriad — APPROVED** [review]
you did it 👍

**Comment by aduth** [comment]
Noting that many of our dependencies define a React peer in the 15.x line, so warnings are logged when installing a new version of a package:

```
npm WARN react-autosize-textarea@0.4.8 requires a peer of react@^0.14.0 || ^15.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-autosize-textarea@0.4.8 requires a peer of react-dom@^0.14.0 || ^15.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-datepicker@0.46.0 requires a peer of react@^0.14.0 || ^15.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-datepicker@0.46.0 requires a peer of react-dom@^0.14.0 || ^15.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-slot-fill@1.0.1 requires a peer of react@^15.5.8 but none is installed. You must install peer dependencies yourself.
npm WARN react-transition-group@1.2.0 requires a peer of react@^15.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-transition-group@1.2.0 requires a peer of react-dom@^15.0.0 but none is installed. You must install peer dependencies yourself.
```

I imagine at least some of these should have updated versions available which define React 16 as a valid peer.

**Comment by gziolo** [comment]
Yes, we can upgrade them after release v1.5. We are using the latest React and React DOM for some time with the bundled code so it seemed like an easier migration path. I can open a follow-up PR on Monday. Thanks for catching this. @youknowriad left a note about this in the preceding PR but I forget to copy it.

**Pull request merged**